test/disk_uploader_test.go

```diff
@@ -0,0 +1,150 @@
+	"context"
+	"os"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeclientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/utils/ptr"
+
+	v1 "kubevirt.io/api/core/v1"
+	kubevirtcliv1 "kubevirt.io/client-go/kubecli"
+	cdiv1beta1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/kubevirt/pkg/libvmi"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/kubevirt/kubevirt-tekton-tasks/test/constants"
+	"github.com/kubevirt/kubevirt-tekton-tasks/test/framework"
+	"github.com/kubevirt/kubevirt-tekton-tasks/test/runner"
+	"github.com/kubevirt/kubevirt-tekton-tasks/test/testconfigs"
+	"github.com/kubevirt/kubevirt-tekton-tasks/test/vm"
+)
+
+var (
+	secretName       string
+	imageDestination string
+	registryKeyId    string
+	registryKey      string
+)
+
+var _ = Describe("Run disk-uploader", func() {
+	f := framework.NewFramework()
+
+	BeforeEach(func() {
+		if f.TestOptions.SkipDiskUploaderTests {
+			Skip("skipDiskUploaderTests is set to true, skipping tests")
+		}
+
+		imageDestination = os.Getenv("IMAGE_DESTINATION")
+		Expect(imageDestination).ToNot(BeEmpty())
+
+		registryKeyId = os.Getenv("REGISTRY_ACCESS_KEY_ID")
+		Expect(registryKeyId).ToNot(BeEmpty())
+
+		registryKey = os.Getenv("REGISTRY_SECRET_KEY")
+		Expect(registryKey).ToNot(BeEmpty())
+
+		secretName = constants.E2ETestsRandomName("disk-uploader-credentials")
+		_, err := createRegistryCredentials(f.CoreV1Client, secretName, f.DeployNamespace)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	It("Extracts disk from VM and upload to container registry", func() {
+		cirrosDataVolume := newCirrosDataVolume()
+		cirrosVm, err := createCirrosVM(f.KubevirtClient, f.DeployNamespace, cirrosDataVolume)
+		Expect(err).ToNot(HaveOccurred())
+
+		f.ManageVMs(cirrosVm)
+		f.ManageDataVolumes(cirrosDataVolume)
+
+		_, err = vm.WaitForVM(f.KubevirtClient, f.DeployNamespace, cirrosVm.Name, "", constants.Timeouts.WaitForVMStart.Duration, false)
+		Expect(err).ToNot(HaveOccurred())
+
+		config := &testconfigs.DiskUploaderTestConfig{
+			TaskRunTestConfig: testconfigs.TaskRunTestConfig{},
+			TaskData: testconfigs.DiskUploaderTaskData{
+				ExportSourceKind: "vm",
+				ExportSourceName: cirrosVm.Name,
+				VolumeName:       cirrosVm.Spec.Template.Spec.Volumes[0].DataVolume.Name,
+				ImageDestination: imageDestination,
+				SecretName:       secretName,
+			},
+		}
+		f.TestSetup(config)
+
+		runner.NewTaskRunRunner(f, config.GetTaskRun()).
+			CreateTaskRun().
+			ExpectSuccess().
+			ExpectLogs(config.GetAllExpectedLogs()...).
+			ExpectResults(nil)
+
+		ref, err := name.ParseReference(imageDestination)
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = remote.Get(ref, remote.WithAuth(&authn.Basic{
```

Comment:
Does the task output an image id or hash? Maybe we can assert that way that the image was updated on the registry?

Reply:
The task does not write any result (because `crane.Push` returns only error and nothing else).  Hower we might take and return the image digest from the variable `containerImage` in this step: https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/cmd/disk-uploader/main.go#L98 (I am not sure if it is populated during build - needs to be tested if it is not empty) and then check it
```
descriptor, err = remote.Get(ref, remote.WithAuth(&authn.Basic{
...
image := descriptor.Image()
Expect(image.Digest).To(Equal(result.Digest))
```

Reply:
> Does the task output an image id or hash? Maybe we can assert that way that the image was updated on the registry?

@0xFelix I'm not sure again if the assertion is needed. When you push an image using `go-containerreigstry` it checks if the image was pushed by checking the response status [1]. It will return an error if it failed to push an image to the container registry.

[1] https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/write.go#L590

> The task does not write any result (because crane.Push returns only error and nothing else).

@ksimon1 I don't understand how the go code inside disk-uploader could update the Tekton Task's result with the image digest, do you have an example for it? besides collecting logs.

Anyways, in your code's example can be added as an extra step after completing the upload of the image to the container registry. This way we can also fail the task's execution from inside execution here [1].

[1] https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/cmd/disk-uploader/main.go#L109

Reply:
Please check how other tasks are returning results: https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/cmd/create-vm/main.go#L59

Reply:
@ksimon1 I'll take a look.

Reply:
Done.